File: structy/graphs.js
const depthFirstPrint = (graph, source) => {
    const stack = [ source ];

    while (stack.length > 0) {
        const current = stack.pop();
        console.log(current);
        for (let neighbor of graph[current]) {
            stack.push(neighbor)
        }
    }
}

const depthFirstPrintRec = (graph, current) => {
    console.log(current)
    for (let neighbor of graph[current]) {
        depthFirstPrintRec(graph, neighbor)
    }
}

const breadthFirstPrint = (graph, start) => {
    const queue = [ start ];

    while (queue.length > 0) {
        const current = queue.shift()
        console.log(current)
        for (let neighbor of graph[current]) {
            queue.push(neighbor)
        }
    }
}

//depthFirstPrint(graph, 'a') // abdfce
//depthFirstPrintRec(graph, 'a')
//breadthFirstPrint(graph, "a")

// hasPath - My Solve (stack)
const hasPathMySolve = (graph, src, dst) => {
    const stack = [ src ]
    
    while (stack.length > 0) {
      const current = stack.pop()
      if (current === dst) return true

      for (let neighbor of graph[current]) {
        stack.push(neighbor)
      }
    }
    return false
  };

// My solve recurisve
const hasPathRec = (graph, src, dst) => {
    if (src === dst) return true
    
    for (let neighbor of graph[src]) {
      if (hasPath(graph, neighbor, dst) === true) {
        return true
      }
    }
    return false 
  };

// My solve queue
const hasPathBFS = (graph, src, dst) => {
    const queue = [ src ]
    
    while (queue.length > 0) {
      const current = queue.shift()
      if (current === dst) return true
      
      for (let neighbor of graph[current]) {
        queue.push(neighbor)
      }
    }
    return false
  };


// Undirected Path
const undirectedPath = (edges, nodeA, nodeB) => {
    const graph = buildGraph(edges)
    const visited = {}
    return hasPath(graph, nodeA, nodeB, new Set())
  };
  
  const hasPath = (graph, src, dst, visited) => {
    if (src === dst) return true
    if (visited.has(src)) return false
    visited.add(src)
    
    for (let neighbor of graph[src]) {
      if (hasPath(graph, neighbor, dst, visited) === true) return true
    }
    return false
  } 
  
  const buildGraph = (edges) => {
    const graph = {}
    
    for (let edge of edges) {
        const [a, b] = edge
        if (!(graph[a])) graph[a] = []
        if (!(graph[b])) graph[b] = []
        graph[a].push(b)
        graph[b].push(a)
    }
    
    return graph
  }

// n = number of nodes 
// e = number of edges
// Time: O(e)
// Space: O(e)

// DFS Rec

const graph = {
    a: ['b', 'c'],
    b: ['d'],
    c: ['e'],
    d: ['f'],
    e: [],
    f: []
}

const graphDFSRecursive = (graph, current) => {
    console.log(current)
    for (let neighbor of graph[current]) {
        graphDFSRecursive(graph, neighbor)
    }
}

console.log(graphDFSRecursive(graph, "a"))

// Connected Components Count
const connectedComponentsCount = (graph) => {
    const visited = new Set()
    let count = 0; 
    
    for (let node in graph) {
      if (explore(graph, node, visited) === true) {
        count += 1
      }
    } 
    return count;
  }
  
  const explore = (graph, current, visited) => {
    if (visited.has(String(current))) return false; 
    
    visited.add(String(current));
    
    for (let neighbor of graph[current]) {
      explore(graph, neighbor, visited)
    }
    return true
  }

// Shortest Path 
const shortestPath = (edges, nodeA, nodeB) => {
    const graph = adjList(edges);
    const queue = [ [nodeA, 0] ];
    const visited = new Set()

    while (queue.length > 0) {
        const [node, level] = queue.shift();
        if (node === nodeB) return level; 

        for (let neighbor of graph[node]) {
            if (!visited.has(neighbor)) {
                visited.add(neighbor)
                queue.push([neighbor, level + 1])
            }
        }
    } 
    return -1
  };

  const adjList = (edges) => {
    const final = {}
    
    for (let edge of edges) {
        const [node1, node2] = edge
        if (!(final[node1])) {
            final[node1] = []
        }
        if (!(final[node2])) {
            final[node2] = []
        }
        final[node1].push(node2)
        final[node2].push(node1)
    }    
    return final
  }
  
edges = [
  ['w', 'x'],
  ['x', 'y'],
  ['z', 'y'],
  ['z', 'v'],
  ['w', 'v']
]

console.log('Shortest Path: ' + shortestPath(edges, 'w', 'z'))

const islandCount = (grid) => {
    let islands = 0;
    const visited = new Set();

    for (let r = 0; r < grid.length; r += 1) {
        for (let c = 0; c < grid[0].length; c+= 1) {
            if (exploreIsl(grid, r, c, visited) === true) {
                islands += 1;
            }
        }
    }
return islands;
};
  
const exploreIsl = (grid, r, c, visited) => {
    const rowInbounds = 0 <= r && r < grid.length;
    const colInbounds = 0 <= c && c < grid[0].length; 

    if (!rowInbounds || !colInbounds) return false; 

    if (grid[r][c] === 'W') return false;

    const pos = r + ',' + c;
    if (visited.has(pos)) return false; 
    visited.add(pos);

    explore(grid, r-1, c, visited);
    explore(grid, r+1, c, visited);
    explore(grid, r, c-1, visited);
    explore(grid, r, c+1, visited);

    return true; 
}

const minimumIsland = (grid) => {
    let minimum = Infinity;
    const visited = new Set(); 

    for (let r = 0; r < grid.length; r+=1 ) {
        for (let c = 0; c < grid[0].length; c+=1) {
            const currentSize = findSize(grid, r, c, visited)
            if (currentSize > 0 && currentSize < minimum) {
                minimum = currentSize
            };
        }
    }
    return minimum;
}

const findSize = (grid, r, c, visited) => {
    const inRow = 0 <= r && r < grid.length; 
    const inCol = 0 <= c && c < grid[0].length;
    if (!inRow || !inCol) return 0; 
    
    if (grid[r][c] === 'W') return 0;

    const pos = r + ',' + c;
    if (visited.has(pos)) return 0; 
    visited.add(pos);

    let size = 1;
    size += findSize(grid, r-1, c, visited)
    size += findSize(grid, r+1, c, visited)
    size += findSize(grid, r, c-1, visited)
    size += findSize(grid, r, c+1, visited)
    return size; 
}

grid = [
  ['W', 'L', 'W', 'W', 'W'],
  ['W', 'L', 'W', 'W', 'W'],
  ['W', 'W', 'W', 'L', 'W'],
  ['W', 'W', 'L', 'L', 'W'],
  ['L', 'W', 'W', 'L', 'L'],
  ['L', 'L', 'W', 'W', 'W'],
];

console.log(minimumIsland(grid)) // 2

const closestCarrot = (grid, startRow, startCol) => {
    shortestPath = -1;
    let pos = [startRow, startCol]
    const queue = [ pos ]    
}

File: python/graphs/graphs.py
from collections import deque

def depth_first_print(graph, start):
    stack = [ start ]
    
    while len(stack) > 0: 
        current = stack[-1]
        print(current)
        stack.pop()
        for neighbor in graph[current]:
            stack.append(neighbor)

def depth_first_print_rec(graph, current):
    print(current)
    for neighbor in graph[current]: 
        depth_first_print_rec(graph, neighbor)
        
def breadth_first_print(graph, start):
    queue = [ start ] 
    
    while queue: 
        current = queue.pop(0) 
        print(current) #
        for neighbor in graph[current]:
            queue.append(neighbor) 

def breadth_first_print_deque(graph, start):
    queue = deque([ start ])
    while queue:
        current = queue[0]
        print(current)
        queue.popleft()
        for neighbor in graph[current]:
            queue.append(neighbor)
    
graph = {
    "a": ["b", "c"], 
    "b": ["d"],
    "c": ["e"],
    "d": ["f"], 
    "e": [],
    "f": []
}

#print(depth_first_print_rec(graph, "a"))
#print(breadth_first_print(graph, "a"))
print(breadth_first_print_deque(graph, "a"))

# has_path(recursive)
def has_path_rec(graph, src, dst):
  if src == dst: return True
  for neighbor in graph[src]:
    if has_path_rec(graph, neighbor, dst):
      return True
  return False 

# has_path BFS 
def has_path_bfs(graph, src, dst):
  queue = deque([src])
  
  while queue:
    current = queue.popleft()
    if current == dst: return True
    for neighbor in graph[current]:
      queue.append(neighbor)
  
  return False 

# n = number of nodes
# e = number of edges
# Time: O(e)
# Space: O(n) 

# undirected path
def undirected_path(edges, node_A, node_B):
    graph = build_graph(edges)
    return has_path(graph, node_A, node_B, set())
  
def has_path(graph, src, dst, visited):
  if src == dst: return True

  if src in visited: 
    return False
  
  visited.add(src)

  for neighbor in graph[src]:
    if has_path(graph, neighbor, dst, visited) == True:
      return True
  
  return False 
  
def build_graph(edges):
  graph = {}
  
  for edge in edges: 
    a, b = edge
    if a not in graph: 
      graph[a] = []
    if b not in graph: 
      graph[b] = []
      
    graph[a].append(b)
    graph[b].append(a)
  
  return graph

# Connected Components Count 
def connectedComponentsCount(graph):
    count = 0
    visited = set()
    
    for node in graph:
        if explore(graph, node, visited) == True:
            count += 1
    
    return count

def explore(graph, current, visited):
    if current in visited:
        return False
    
    visited.add(current)
    
    for neighbor in graph[current]:
        explore(graph, neighbor, visited)
    
    return True

print(connectedComponentsCount(graph))

def shortest_path(edges, node_a, node_b):
    graph = _create_graph(edges)
    visited = set()
    queue = [ [node_a, 0] ]
    
    while len(queue) > 0:
        [node, level] = queue.pop(0)
        
        if node == node_b:
            return level
        
        for neighbor in graph[node]:     
            if node not in visited:
                visited.add(node)
                queue.append([neighbor, level + 1])
                
    return -1 

def _create_graph(edges):
    graph = {}
    
    for node in edges: 
        [node_a, node_b] = node
        if node_a not in graph:
            graph[node_a] = []
        if node_b not in graph:
            graph[node_b] = []
    
        if node_a in graph:
            graph[node_a].append(node_b)
        if node_b in graph:
            graph[node_b].append(node_a)
    
    return graph

def shortest_path(edges, node_A, node_B):
    graph = _create_graph2(edges)
    visited = set([ node_A ])
    queue = deque([ (node_A, 0) ]) 
    
    while queue:
        node, level = queue.popleft()
        
        if node == node_B:
            return level
        
        for neighbor in graph[node]:     
            if neighbor not in visited:
                visited.add(node)
                queue.append((neighbor, level + 1))
                
    return -1 

def _create_graph2(edges):
    graph = {}
    
    for node in edges: 
        node_a, node_b = node
        if node_a not in graph:
            graph[node_a] = []
        if node_b not in graph:
            graph[node_b] = []
    
        graph[node_a].append(node_b)
        graph[node_b].append(node_a)
    
    return graph

def island_count(grid):
  visited = set()
  islands = 0 
  
  for row in range(len(grid)):
    for col in range(len(grid[0])):
      if exploreIsland(grid, row, col, visited) == True:
        islands += 1 
  
  return islands

def exploreIsland(grid, r, c, visited):
  inRow =  0 <= r and r < len(grid)  
  inCol =  0 <= c and c < len(grid[0]) 
  if inRow == False or inCol == False: 
    return False
  
  if grid[r][c] == "W": return False

  pos = (r, c)
  if pos in visited: return False
  visited.add(pos)
  
  exploreIsland(grid, r-1, c, visited)
  exploreIsland(grid, r+1, c, visited)
  exploreIsland(grid, r, c-1, visited)
  exploreIsland(grid, r, c+1, visited)
  return True

def minimum_island(grid):
    minimum = float('inf')
    visited = set()
    
    for r in range(len(grid)):
        for c in range(len(grid[0])):
            size = find_size(grid, r, c, visited)
            if size > 0 and size < minimum:
                minimum = size
    
    return minimum

def find_size(grid, r, c, visited):
    inRow = 0 <= r and r < len(grid)
    inCol = 0 <= c and c < len(grid[0])
    if (inRow == False or inCol == False):
        return 0
    
    if grid[r][c] == 'W':
        return 0
    
    pos = (r, c)
    if pos in visited:
        return 0
    visited.add(pos)
    
    size = 1
    size += find_size(grid, r - 1, c, visited)
    size += find_size(grid, r + 1, c, visited)
    size += find_size(grid, r, c - 1, visited)
    size += find_size(grid, r, c + 1, visited)
    
    return size

grid = [
  ['W', 'L', 'W', 'W', 'W'],
  ['W', 'L', 'W', 'W', 'W'],
  ['W', 'W', 'W', 'L', 'W'],
  ['W', 'W', 'L', 'L', 'W'],
  ['L', 'W', 'W', 'L', 'L'],
  ['L', 'L', 'W', 'W', 'W'],
]

print(minimum_island(grid))